test/external-modules/latency-simulating-directory/src/main/java/org/elasticsearch/test/inputlatency/SimulatedLatencyDirectoryWrapper.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.test.inputlatency;
+
+import org.apache.lucene.store.BufferedIndexInput;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FilterDirectory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.NIOFSDirectory;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.core.IOUtils;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.index.IndexModule;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.function.Consumer;
+
+/**
+ * A DirectoryWrapper that sleeps for 50ms whenever an indexinput fills its internal
+ * data buffer
+ */
+public class SimulatedLatencyDirectoryWrapper implements IndexModule.DirectoryWrapper {
+
+    private final Consumer<String> delayer;
+
+    public SimulatedLatencyDirectoryWrapper(Consumer<String> delayer) {
+        this.delayer = delayer;
+    }
+
+    @Override
+    public Directory wrap(Directory directory, ShardRouting shardRouting) {
+        if (directory instanceof NIOFSDirectory niofsDir) {
+            return new FilterDirectory(niofsDir) {
+                @Override
+                public IndexInput openInput(String name, IOContext context) throws IOException {
+                    Path path = niofsDir.getDirectory().resolve(name);
+                    FileChannel fc = FileChannel.open(path, StandardOpenOption.READ);
+                    boolean success = false;
+                    try {
+                        final DelayingNIOFSIndexInput indexInput = new DelayingNIOFSIndexInput(
+                            "NIOFSIndexInput(path=\"" + path + "\")",
+                            fc,
+                            context
+                        );
+                        success = true;
+                        return indexInput;
+                    } finally {
+                        if (success == false) {
+                            IOUtils.closeWhileHandlingException(fc);
+                        }
+                    }
+                }
+            };
+        } else {
+            throw new IllegalStateException("Can only wrap NIOFSDirectory");
+        }
+    }
+
+    @SuppressForbidden(reason = "Copied from lucene")
+    final class DelayingNIOFSIndexInput extends BufferedIndexInput {
```

Comment:
(I don't have all the context so sorry if my comment is out of the scope)

Instead of implementing our own BufferedIndexInput, could it be a `FilterIndexInput` that count the bytes read and maybe pause/sleep if needed? That would prevent to copy over all the logic of reading from a FileChannel.

If we intend to simulate delays like the frozen cache, delaying on every buffer refill is a bit extreme: delaying on the first read and then every 16mb might be more close to how the frozen cache works I think.